### PR TITLE
[Messenger] Add Beanstalkd bridge

### PR DESCRIPTION
**Pull request by X-Coder264**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       |
| License       | MIT
| Doc PR        |

On a project at work we use Beanstalkd and we wanted to start using Messenger with it so I've written this bridge/transport. Any feedback is highly appreciated.

### CODE REVIEW

File: src/Symfony/Component/Messenger/Bridge/Beanstalkd/Transport/BeanstalkdSender.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+/**
+ * @author Antonio Pauletich <antonio.pauletich95@gmail.com>
+ */
+class BeanstalkdSender implements SenderInterface
+{
+    private $connection;
+    private $serializer;
+
+    public function __construct(Connection $connection, SerializerInterface $serializer)
```
Comment (noniagriconomie) [inline src/Symfony/Component/Messenger/Bridge/Beanstalkd/Transport/BeanstalkdSender.php:27]: Serializer can be nullable here no?

Comment (X-Coder264) [reply src/Symfony/Component/Messenger/Bridge/Beanstalkd/Transport/BeanstalkdSender.php]: Fixed.

**Comment by X-Coder264** [comment]
PR rebased with master.

I don't understand why the Travis PHP 7.2 build fails with the message `PHP Fatal error:  Class 'Pheanstalk\Contract\PheanstalkInterface' not found in /home/travis/build/symfony/symfony/src/Symfony/Component/Messenger/Bridge/Beanstalkd/Tests/Transport/ConnectionTest.php on line 42` even though Pheanstalk is required by the package. The 7.3 and 7.4 jobs are passing.

File: src/Symfony/Component/Messenger/composer.json [hunk 2]
```diff
@@ -19,6 +19,7 @@
         "php": ">=7.2.5",
         "psr/log": "~1.0",
         "symfony/amqp-messenger": "^5.1",
+        "symfony/beanstalkd-messenger": "^5.2",
```
Comment (fabpot) [inline src/Symfony/Component/Messenger/composer.json:22]: Should be removed. The idea is to add it only when needed.We do have AMQP for BC reasons.

**Review by fabpot — CHANGES_REQUESTED** [review]
Haven't had time to do a full review. Just spotted one change for now.

File: src/Symfony/Component/Messenger/composer.json (hunk 2 above)
Comment (X-Coder264) [reply src/Symfony/Component/Messenger/composer.json]: Oh right, removed.

File: src/Symfony/Component/Messenger/Bridge/Beanstalkd/README.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
+Beanstalkd Messenger
+===============
```
Comment (fabpot) [inline src/Symfony/Component/Messenger/Bridge/Beanstalkd/README.md:2]: Missing some `=` chars

File: src/Symfony/Component/Messenger/Bridge/Beanstalkd/README.md [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+Provides Beanstalkd integration for Symfony Messenger.
+
```
Comment (fabpot) [inline src/Symfony/Component/Messenger/Bridge/Beanstalkd/README.md:5]: Can you add the DSN pattern?

File: src/Symfony/Component/Messenger/Bridge/Beanstalkd/Transport/Connection.php [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+    /**
+     * Available options:
+     *
+     * * tube_name: name of the tube
+     * * timeout: message reservation timeout (in seconds)
+     * * ttr: the message time to run before it is put back in the ready queue (in seconds)
+     */
+    private $configuration;
+
```
Comment (fabpot) [inline src/Symfony/Component/Messenger/Bridge/Beanstalkd/Transport/Connection.php:44]: blank line can be removed

**Review by fabpot — APPROVED** [review]
LGTM (with some very minor comments)

File: src/Symfony/Component/Messenger/Bridge/Beanstalkd/README.md (hunk 3 above)
Comment (X-Coder264) [reply src/Symfony/Component/Messenger/Bridge/Beanstalkd/README.md]: I've added the missing `=` chars.

File: src/Symfony/Component/Messenger/Bridge/Beanstalkd/README.md (hunk 4 above)
Comment (X-Coder264) [reply src/Symfony/Component/Messenger/Bridge/Beanstalkd/README.md:5]: I've added the DSN "pattern". There aren't any in the other bridges in the README so hopefully it's ok how I've written the pattern.

File: src/Symfony/Component/Messenger/Bridge/Beanstalkd/Transport/Connection.php (hunk 5 above)
Comment (X-Coder264) [reply src/Symfony/Component/Messenger/Bridge/Beanstalkd/Transport/Connection.php]: Removed.

**Review by fabpot — APPROVED** [review]

**Review by fabpot — CHANGES_REQUESTED** [review]
@X-Coder264 Can you work on fixing the test now? I suppose that we need to install Beanstalkd on Travis.

**Comment by X-Coder264** [comment]
@fabpot You mean this failure? https://travis-ci.org/github/symfony/symfony/jobs/717325508
This happens only on the PHP 7.2 build, but I don't understand why since the Pheanstalk library is required by the bridge.
On the PHP 7.4 build I can see

```
cd src/Symfony/Component/Messenger/Bridge/Beanstalkd

+'[' -e composer.lock ']'

+composer install --no-progress --no-suggest --ansi

Loading composer repositories with package information

Installing dependencies (including require-dev) from lock file

Package operations: 7 installs, 0 updates, 0 removals

  - Installing pda/pheanstalk (v4.0.0): Downloading (100%)

  - Installing psr/log (1.0.0): Loading from cache

  - Installing symfony/messenger (v4.4.0): Loading from cache

  - Installing symfony/inflector (v3.4.0): Loading from cache

  - Installing symfony/property-access (v4.4.0): Loading from cache

  - Installing symfony/polyfill-ctype (v1.8.0): Loading from cache

  - Installing symfony/serializer (v4.4.0): Loading from cache

+/home/travis/build/symfony/symfony/phpunit --exclude-group tty,benchmark,intl-data
```

The PHP 7.2 build does not do that, it just runs

```
/home/travis/build/symfony/symfony/phpunit --exclude-group tty,benchmark,intl-data src/Symfony/Component/Messenger/Bridge/Beanstalkd
```

I'm not familiar with your Travis CI setup so I'll have to dig a bit deeper into it to check why `composer install` doesn't happen on the 7.2 build (from the look of it it seems like the 7.2 build is triggering some other code path in `.travis.yml`).

**Comment by fabpot** [comment]
@X-Coder264 You need to add `"pda/pheanstalk": "^4.0",` in the `require-dev` section of the main/root `composer.json`.

**Comment by X-Coder264** [comment]
> @X-Coder264 You need to add `"pda/pheanstalk": "^4.0",` in the `require-dev` section of the main/root `composer.json`.

Thanks for the tip - I've added it.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @X-Coder264.

**Comment by X-Coder264** [comment]
@fabpot Thanks for the merge. I'll create a PR for the documentation this weekend.

**Pull request merged**